### PR TITLE
Feature/answers

### PR DESCRIPTION
**Pull request by ppauljonathan**
Made users able answer questions and displayed answers on the same page and also made CRUD of answers

### CODE REVIEW

**Review by shubhamjuneja1 — CHANGES_REQUESTED** [review]

File: app/controllers/users_controller.rb [hunk 1]
```diff
@@ -46,10 +49,15 @@ def update
...
 
   private def set_user
-    @user = User.includes(:topics).find_by_id(params[:id])
+    @user = User.with_topics.with_profile_picture.find_by_id(params[:id])
     redirect_to root_path, alert: 'user not found' unless @user
   end
 
+  private def set_answers
+    @answers = @user.answers.with_content
+    @answers = @answers.published unless @user == current_user
```
[the remark this answers is not in the document]
Comment (ppauljonathan) [reply app/controllers/users_controller.rb]: removed

File: app/controllers/users_controller.rb [hunk 2]
```diff
@@ -1,10 +1,13 @@
...
-
   before_action :set_user, except: %i[index]
   before_action :check_if_current_user, only: %i[edit update drafts destroy]
+  before_action :set_answers, only: :answers
+
+  skip_before_action :authorize, only: %i[index show questions answers comments]
 
-  skip_before_action :authorize, only: %i[index show]
+  def answers
+    render :show
```
[the remark this answers is not in the document]
Comment (ppauljonathan) [reply app/controllers/users_controller.rb]: removed

File: app/models/user.rb [hunk 3]
```diff
@@ -16,11 +16,15 @@ class User < ApplicationRecord
 
   has_one_attached :profile_picture
   has_many :questions, dependent: :nullify
+  has_many :answers, dependent: :nullify
 
   acts_as_taggable_on :topics
 
   enum :role, ROLES, default: :user
 
+  scope :with_topics, -> { includes :topics }
```
[the remark this answers is not in the document]
Comment (ppauljonathan) [reply app/models/user.rb]: removed

File: app/views/questions/show.html.erb [hunk 4]
```diff
@@ -1,51 +1,9 @@
...
-  <% @question.files_attachments.each do |file| %>
-    <div class="file-item">
-      <%= image_tag file.preview(resize_to_limit: [100, 100]) %>
-      <br>
-      <%= file.blob.filename %>
-      <%= link_to 'Download', rails_blob_url(file) %>
-    </div>
-  <% end %>
-</div>
+<% if @question.published_at? %>
```
Comment (shubhamjuneja1) [inline app/views/questions/show.html.erb:3]: why do we need this check?
users will only this when question is published
question's owner can see this, no harm

File: app/views/questions/_question_file_attachments.html.erb [hunk 5]
```diff
@@ -0,0 +1,13 @@
+<% if question.files.any? %>
+  <div class="form-heading" style="margin-left: 4%">Attached Files:</div>
+  <div class="question-files" style="margin-left: 4%">
+    <% @question.files_attachments.each do |file| %>
```
Comment (shubhamjuneja1) [inline app/views/questions/_question_file_attachments.html.erb:4]: have you eager loaded files_attachments?

File: app/controllers/questions_controller.rb [hunk 6]
```diff
@@ -26,7 +29,10 @@ def destroy
   end
 
   def index
+    @paginated_questions = Question.page(params[:page]).per(QUESTIONS_PER_PAGE)
     @search_results = Question.published
+                              .page(params[:page])
+                              .per(QUESTIONS_PER_PAGE)
```
Comment (shubhamjuneja1) [inline app/controllers/questions_controller.rb:35]: have you checked kaminari?
there is a default for this, which you can override in models

**Review by shubhamjuneja1 — CHANGES_REQUESTED** [review]

File: app/views/questions/show.html.erb (hunk 4 above)
Comment (ppauljonathan) [reply app/views/questions/show.html.erb:3]: removed form here

File: app/views/questions/_question_file_attachments.html.erb (hunk 5 above)
Comment (ppauljonathan) [reply app/views/questions/_question_file_attachments.html.erb:4]:
```ruby
private def set_question
    @question = Question.includes(:topics,
                                  :files_attachments,
                                  :rich_text_content,
                                  { user: :profile_picture_attachment },
                                  { answers: :rich_text_content })
                        .find_by_url_slug(params[:url_slug])
  end
```
yes

File: app/controllers/questions_controller.rb (hunk 6 above)
Comment (ppauljonathan) [reply app/controllers/questions_controller.rb:35]: resolved in next PR #5
[here](https://github.com/ppauljonathan/quora-clone/blob/1ba328c46dd6dd8638178a564f290576e4de797c/app/controllers/questions_controller.rb#L34l)

**Review by shubhamjuneja1 — APPROVED** [review]

**Pull request merged**